internals/overlord/logstate/manager.go

```diff
@@ -15,26 +15,98 @@
 package logstate
 
 import (
+	"sync"
+
+	"github.com/canonical/pebble/internals/logger"
 	"github.com/canonical/pebble/internals/plan"
 	"github.com/canonical/pebble/internals/servicelog"
 )
 
-type LogManager struct{}
+type LogManager struct {
+	forwarders map[string]*logForwarder
+	gatherers  map[string]*logGatherer
+
+	newForwarder func(serviceName string) *logForwarder
+	newGatherer  func(*plan.LogTarget) *logGatherer
+}
 
 func NewLogManager() *LogManager {
-	return &LogManager{}
+	return &LogManager{
+		forwarders:   map[string]*logForwarder{},
+		gatherers:    map[string]*logGatherer{},
+		newForwarder: newLogForwarder,
+		newGatherer:  newLogGatherer,
+	}
 }
 
-// PlanChanged is called by the service manager when the plan changes. We stop
-// all running forwarders, and start new forwarders based on the new plan.
+// PlanChanged is called by the service manager when the plan changes. We update the list of gatherers for each forwarder based on the new plan.
 func (m *LogManager) PlanChanged(pl *plan.Plan) {
-	// TODO: implement
+	// Create a map to hold forwarders/gatherers for the new plan.
+	// Old forwarders/gatherers will be moved over or deleted.
+	newForwarders := make(map[string]*logForwarder, len(pl.Services))
+	newGatherers := make(map[string]*logGatherer, len(pl.LogTargets))
+
+	for serviceName, service := range pl.Services {
+		// TODO: don't create forwarders if there are no targets for this service?
+		forwarder := m.forwarders[serviceName]
+		if forwarder == nil {
+			// Create new forwarder
+			forwarder = m.newForwarder(serviceName)
+			newForwarders[serviceName] = forwarder
+		} else {
+			// Copy over existing forwarder
+			newForwarders[serviceName] = forwarder
+			delete(m.forwarders, serviceName)
+		}
+
+		// update clients
+		forwarder.mu.Lock()
+		forwarder.gatherers = []*logGatherer{}
+
+		for _, target := range pl.LogTargets {
+			// Only create the gatherer if there is a service logging to it.
+			if service.LogsTo(target) {
```

Comment:
invert this if statement and continue to avoid indentation.
```
if !service.LogsTo(target) {
    continue
}
gatherer := m.gatherers[serviceName]
...